File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
  def isValid(self, s: str) -> bool:
    stack = []
    open = {
      ')': '(',
      ']': '[',
      '}': '{',
    }
    for p in s:
      if p in '([{':
        stack.append(p)
      elif len(stack) == 0 or stack.pop(-1) != open[p]:
        return False
    return not len(stack)

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
  public boolean isValid(String s) {
    List<String> stack = new ArrayList<String>();
    Map open = new HashMap();
    open.put(")", "(");
    open.put("]", "[");
    open.put("}", "{");
    
    for (int i = 0; i < s.length(); i++) {
      String p = s.substring(i, i + 1);
      if (open.containsValue(p)) {
        stack.add(p);
      } else {
        if (stack.isEmpty()) return false;
        String popped = stack.get(stack.size() - 1);
        if (!popped.equals(open.get(p))) {
          return false;
        } else {
          stack.remove(stack.size() - 1);
        }
      }
    }
    return stack.isEmpty();
  }
}